### DISCUSSION TITLE
RKE2 cluster upgrade, can I upgrade the Worker Node (WN) one by one?

### DISCUSSION THREAD

**Discussion by nobody2749023804** [rancher/rke2#5256] [Q&A] [answered]
Not sure if this is the right place to ask this question. I used to ask at Rancher Slack. But yesterday I found that I suddenly was not allowed to post message over there. Myabe my Slack account problem.

Back to the question, as below.

We have a RKE2 1.22 cluster and want to upgrade to 1.23. (We know 1.23 is also old, we do hope to further upgrade after then.)

We see that Rancher has built-in upgrade function and seems pretty simple and straightforward.
https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/upgrade-and-roll-back-kubernetes

However, the upgrade function seems to upgrade all Worker Node (WN) in one go?
I mean, although it will upgrade the WNs one by one, or some by some depending on your "Maximum Unavailable Worker Nodes" setting, but it seems to keep upgrading the WNs some by some non-stop until all are upgraded?

To play safe, and in a more controlled manner of risk and possible downtime, can we break down the whole upgrade process into multiple days, or even few weeks?
Say, upgrade 1 WN first, if OK, then upgrade another next week, if OK, then so and so on? and we get to choose (control) which WN to upgrade each time?
Which leads to another question: in a RKE2 cluster, can different WNs running different versions of RKE2? e.g. if we upgrade 1 WN only each time, then during the transition period some WNs will be running RKE2 1.22 while some 1.23

Any Rancher doc or guidelines about these?

**Comment by manuelbuil** [comment, marked as answer]
You could use https://docs.rke2.io/upgrade/automated_upgrade. In the plan, you can specify a label (as in the example):
```
      - {key: rke2-upgrade, operator: Exists}
      - {key: rke2-upgrade, operator: NotIn, values: ["disabled", "false"]}
```
And then set that label only in the nodes that you want to upgrade

**Reply by nobody2749023804** [reply]
Thanks. If this feature (upgrade some nodes only, not all) is officially supported, does that mean nodes of different RKE2 versions (in our case, 1.22 and 1.23) can co-exist in one single RKE2 cluster, all normally running workloads, not cordoned (disabled)?

**Reply by manuelbuil** [reply]
I think it should work but I would not recommend it for a long time. In the end it is different code and there might be unexpected issues. In fact, I don't think we currently have any test verifying this scenario. Anyway, this is open source, so to state that an specific scenario is "officially supported" is complicated :)

**Reply by nobody2749023804** [reply]
Noted and thanks~

**Comment by manuelbuil** [comment]
BTW, the rancher user slack was closed some hours because there was some sort of crypto spam being sent uncontrollably. Probably that's the reason you experienced problems :)

**Reply by nobody2749023804** [reply]
Thanks. Good to know that it's not my slack account's problem :)
